file_store/src/file_info_poller.rs

```diff
@@ -227,91 +265,92 @@ fn create_cache() -> MemoryFileCache {
 }
 
 async fn is_already_processed(
-    db: impl sqlx::PgExecutor<'_>,
+    state: &impl FileInfoPollerState,
     cache: &MemoryFileCache,
+    process_name: &str,
     file_info: &FileInfo,
 ) -> Result<bool> {
     if cache.get(&file_info.key).await.is_some() {
         Ok(true)
     } else {
-        db::exists(db, file_info).await
+        state.exists(process_name, file_info).await
     }
 }
 
 async fn cache_file(cache: &MemoryFileCache, file_info: &FileInfo) {
     cache.insert(file_info.key.clone(), true, CACHE_TTL).await;
 }
 
-mod db {
-    use super::*;
+#[cfg(feature = "sqlx-postgres")]
+#[async_trait::async_trait]
+impl FileInfoPollerStateRecorder for &mut sqlx::Transaction<'_, sqlx::Postgres> {
+    async fn record(self, process_name: &str, file_info: &FileInfo) -> Result {
+        sqlx::query(
+            r#"
+                INSERT INTO files_processed(process_name, file_name, file_type, file_timestamp, processed_at) VALUES($1, $2, $3, $4, $5)
+            "#)
+            .bind(process_name)
+            .bind(&file_info.key)
+            .bind(&file_info.prefix)
+            .bind(file_info.timestamp)
+            .bind(Utc::now())
+            .execute(self)
+            .await
+            .map(|_| ())
+            .map_err(Error::from)
+    }
+}
 
-    pub async fn latest_ts(
-        db: impl sqlx::PgExecutor<'_>,
+#[cfg(feature = "sqlx-postgres")]
+#[async_trait::async_trait]
+impl FileInfoPollerState for sqlx::Pool<sqlx::Postgres> {
+    async fn latest_timestamp(
+        &self,
+        process_name: &str,
         file_type: &str,
     ) -> Result<Option<DateTime<Utc>>> {
-        let default = Utc.timestamp_opt(0, 0).single().unwrap();
-
-        let result = sqlx::query_scalar::<_, DateTime<Utc>>(
+        sqlx::query_scalar::<_, Option<DateTime<Utc>>>(
             r#"
-        SELECT COALESCE(MAX(file_timestamp), $1) FROM files_processed where file_type = $2
-        "#,
-        )
-        .bind(default)
-        .bind(file_type)
-        .fetch_one(db)
-        .await?;
-
-        if result == default {
-            Ok(None)
-        } else {
-            Ok(Some(result))
-        }
+                SELECT MAX(file_timestamp) FROM files_processed where process_name = $1 and file_type = $2
+            "#,
+            )
+            .bind(process_name)
+            .bind(file_type)
+            .fetch_one(self)
```

Comment:
could this ever get called with _no_ results such that it needs to be `fetch_optional`?

Reply:
No, when then are no rows that match the where clause, calling the max function will return 1 row where the value is NULL hence mapping the single value to a `<Option<DateTime<Utc>>>`